cli/exec.go

```diff
@@ -85,53 +88,81 @@ func (e *ExecCommand) runExec(args []string) error {
 
 	conn, reader, err := apiClient.ContainerStartExec(ctx, createResp.ID, startExecConfig)
 	if err != nil {
-		return fmt.Errorf("failed to create exec: %v", err)
+		return fmt.Errorf("failed to start exec: %v", err)
 	}
-	defer conn.Close()
 
 	// handle stdio.
-	var wg sync.WaitGroup
-
-	if createExecConfig.AttachStderr || createExecConfig.AttachStdout {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			if !e.Terminal {
-				stdcopy.StdCopy(os.Stdout, os.Stderr, reader)
-			} else {
-				io.Copy(os.Stdout, reader)
+	if err := holdHijackConnection(ctx, conn, reader, createExecConfig.AttachStdin, createExecConfig.AttachStdout, createExecConfig.AttachStderr, e.Terminal); err != nil {
+		return err
+	}
+
+	execInfo, err := apiClient.ContainerExecInspect(ctx, createResp.ID)
+	if err != nil {
+		return err
+	}
+
+	code := execInfo.ExitCode
+	if code != 0 {
+		return ExitError{Code: int(code)}
+	}
+
+	return nil
+}
+
+func holdHijackConnection(ctx context.Context, conn net.Conn, reader *bufio.Reader, stdin, stdout, stderr, tty bool) error {
+	if stdin && tty {
+		in, out, err := setRawMode(true, false)
+		if err != nil {
+			return fmt.Errorf("failed to set raw mode")
+		}
+		defer func() {
+			if err := restoreMode(in, out); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to restore term mode")
 			}
 		}()
 	}
 
-	if createExecConfig.AttachStdin {
-		if e.Terminal {
-			in, out, err := setRawMode(true, false)
-			if err != nil {
-				return fmt.Errorf("failed to set raw mode")
+	stdoutDone := make(chan error, 1)
+	go func() {
+		var err error
+		if stderr || stdout {
+			if !tty {
+				_, err = stdcopy.StdCopy(os.Stdout, os.Stderr, reader)
+			} else {
+				_, err = io.Copy(os.Stdout, reader)
 			}
-			defer func() {
-				if err := restoreMode(in, out); err != nil {
-					fmt.Fprintf(os.Stderr, "failed to restore term mode")
-				}
-			}()
 		}
+		stdoutDone <- err
+	}()
 
-		go func() {
+	stdinDone := make(chan struct{})
+	go func() {
+		if stdin {
 			io.Copy(conn, os.Stdin)
-		}()
-	}
+		}
 
-	wg.Wait()
+		// TODO: close write side of conn
+		close(stdinDone)
+	}()
 
-	execInfo, err := apiClient.ContainerExecInspect(ctx, createResp.ID)
-	if err != nil {
-		return err
-	}
+	select {
+	case err := <-stdoutDone:
+		if err != nil {
+			logrus.Debugf("receive stdout error: %s", err)
+			return err
```

Comment:
If the code goes into this part, the following code will go into the end?

```
go func() {
    if stdin {
       io.Copy(conn, os.Stdin)
    }

    close(stdinDone)
}()
```

Reply:
It is nit comment. If there exists goroutine leak, the cli will die to make the goroutine exit....

Reply:
Maybe depend on when golang schedule the goroutine if stdin is `flase`, if stdin if true, it must became a leak.

Reply:
Yes. But the cli will return and exit so that the leak problem will be solved. LGTM right now.